kafka-connect-s3/src/test/java/io/confluent/connect/s3/integration/S3SinkConnectorNetworkIT.java

```diff
@@ -0,0 +1,333 @@
+  }
+
+  private void alterReadWritePolicyOfBucket(String bucketName) {
+
+    Statement allowRestrictedWriteStatement = new Statement(Statement.Effect.Allow)
+      .withPrincipals(new Principal(System.getenv("SECONDARY_USER_ACCOUNT_ID")))
+      .withActions(S3Actions.GetObject)
+      .withResources(new S3ObjectResource(bucketName, "*"));
+
+    Policy policy = new Policy().withStatements(allowRestrictedWriteStatement);
+
+    S3Client.setBucketPolicy(bucketName, policy.toJson());
+  }
+
+  private int sendRecordsToKafka() {
+    // Send records to Kafka
+    int totalNoOfRecordsProduced = 0;
+    for (int i = 0; i < NUM_RECORDS_PRODUCED; i++) {
+      totalNoOfRecordsProduced++;
+      String kafkaTopic = KAFKA_TOPICS.get(i % KAFKA_TOPICS.size());
+      String kafkaKey = "simple-key-" + i;
+      String kafkaValue = "simple-message-" + i;
+      log.debug("Sending message {} with topic {} to Kafka broker {}", kafkaTopic, kafkaValue);
+      connect.kafka().produce(kafkaTopic, kafkaKey, kafkaValue);
+    }
+    return totalNoOfRecordsProduced;
+  }
+
+  private Map<String, String> getConnectorProps() {
+    Map<String, String> props = new HashMap<>();
+    props.put(SinkConnectorConfig.TOPICS_CONFIG, String.join(",", KAFKA_TOPICS));
+    props.put(CONNECTOR_CLASS_CONFIG, "io.confluent.connect.s3.S3SinkConnector");
+    props.put(TASKS_MAX_CONFIG, Integer.toString(TASKS_MAX));
+
+    props.put(REGION_CONFIG, "ap-south-1");
+    props.put(PART_SIZE_CONFIG, "5242880");
+    props.put(S3_BUCKET_CONFIG, S3_BUCKET);
+    props.put(FLUSH_SIZE_CONFIG , Integer.toString(FLUSH_SIZE));
+    props.put(STORAGE_CLASS_CONFIG, "io.confluent.connect.s3.storage.S3Storage");
+    props.put(PARTITIONER_CLASS_CONFIG, "io.confluent.connect.storage.partitioner.DefaultPartitioner");
+
+    // converters
+    props.put(KEY_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+    props.put(VALUE_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+
+    props.put(FORMAT_CLASS_CONFIG, JSON_FORMAT_CLASS);
+    // license properties
+    return props;
+  }
+
+  private void deleteBucket(String bucketName) {
+    emptyBucket(bucketName);
+    // After all objects are deleted, delete the bucket.
+    S3Client.deleteBucket(bucketName);
+  }
+
+  private void emptyBucket(String bucketName) {
+    // delete all objects to empty bucket
+    ObjectListing objectListing = S3Client.listObjects(bucketName);
+    while (true) {
+      Iterator<S3ObjectSummary> objIter = objectListing.getObjectSummaries().iterator();
+      while (objIter.hasNext()) {
+        S3Client.deleteObject(bucketName, objIter.next().getKey());
+      }
+
+      if (objectListing.isTruncated()) {
+        objectListing = S3Client.listNextBatchOfObjects(objectListing);
+      } else {
+        break;
+      }
+    }
+    // delete versioned objects
+    VersionListing versionList = S3Client.listVersions(new ListVersionsRequest().withBucketName(bucketName));
+    while (true) {
```

Comment:
does the first delete loop leave versioned objects in the bucket?

Reply:
Yes, For versioned buckets, when we attempt to delete the objects, Amazon S3 inserts delete markers for all objects but doesn't delete the object versions. Therefore we need to delete the versions separately.
Link to relevant doc: https://docs.aws.amazon.com/AmazonS3/latest/dev/delete-or-empty-bucket.html#delete-bucket-awssdks